### ISSUE TITLE
When shutting down a silo, new grain is activated on another silo before the old has finished deactivating.

### ISSUE THREAD

**Issue by mikkoairaksinen** [dotnet/orleans#5687]
Setup:

3 silos running on separate EC2 instances in docker
3 frontend servers

Load test client generating a minor load (~100 grains per silo, ~1-2 messages per second)

mobile client sending messages to a grain through a frontend

Grain is activated on Silo1
Silo1 is shut down

Log message timestamps (filtered by the id of this player's grain):

17:15:48.770 silo-2 message indicating the grain has been activated
17:15:49.575 silo-1 message indicating grain is being shut down
17:15:49.585 silo-1 message indicating state is written to the database
17:15:52.831 silo-1 message indicating the graceful shutdown of the silohost has completed

The occurence rate is about 1-2% of grains under mild load. This is obviously a massive issue because the state gets corrupted as the flush in silo-1 doesn't complete before silo-2 claims ownership of the data.

Is there any potential setup error that can cause this?

**Comment by sergeybykov** [comment]
>17:15:49.585 silo-1 message indicating state is written to the database

Does this mean that grain state gets persisted upon deactivation of a grain?

**Comment by mikkoairaksinen** [comment]
> > 17:15:49.585 silo-1 message indicating state is written to the database
> 
> Does this mean that grain state gets persisted upon deactivation of a grain?

Yes, grain state is persisted at fixed intervals with a timer and also at deactivation.

**Comment by sergeybykov** [comment]
We generally recommend not to rely on persiting state upon deactivation because it is not guaranteed to always happen, e.g. in case of a failure.

>the state gets corrupted as the flush in silo-1 doesn't complete before silo-2 claims ownership of the data.

How can it get corrupted? Is persistence of a grain state not an atomic operation? Or do you mean that a new activation of the grain in silo-2 loads state before it gets updated by silo-1?

**Comment by mikkoairaksinen** [comment]
@sergeybykov Yes we have certain checkpoints of critical progress where we always write and understand that during a silo crash for example we will lose some data and that is OK. This use case however would happen whenever we scale down the cluster and since that is very normal operation it would be pretty bad if we consistently lose up to 5 minutes of progress on 1% of players on the silo when it gets shut down.

By corruption I mean that up to 5 minutes (the ticking persist interval) of progress is lost in these scenarios since the silo-2 claims ownership of the optimistic lock to the data upon activation. (We changed this from the default where the lock is claimed on write, but that would be equally bad here since silo-1 would claim the lock right after silo-2 activation and the first write of silo-2 after 5 minutes would then fail)

**Comment by sergeybykov** [comment]
Okay, so this is a data loss, not a corruption, then. That's better.

We've been making improvements to the shutdown sequence recently to reduce the chance of something like this happening. Which version are yo running when you see this behavior?

**Comment by mikkoairaksinen** [comment]
We are using Orleans 2.3.2 currently

**Comment by sergeybykov** [comment]
In theory this shouldn't be happening in case of a graceful shutdown. Assigning to @benjaminpetit to take a look.

**Comment by benjaminpetit** [comment]
Thanks for the detailed description @mikkoairaksinen !

In fact it seems that you point to a real issue:  the `LocalGrainDirectory` will remove entries that point to the silo that is shutting down ([see here]((https://github.com/dotnet/orleans/blob/c88fbf2c8ab5a481d3fda689cc1f729c1bd5fc9d/src/Orleans.Runtime/GrainDirectory/LocalGrainDirectory.cs#L458)

`status.IsTerminating()` returns true if the status is equals to `ShuttingDown`. It think it would be safe to explicitly check the condition `siloStatus == SiloStatus.Stopping || siloStatus == SiloStatus.Dead` in the `LocalGrainDirectory`.

**Comment by mikkoairaksinen** [comment]
Thanks for the fast responses! We'll work around this as much as possible for now until a fix comes in.

**Comment by mikkoairaksinen** [comment]
Any updates on this? It is still causing us issues in production where we can't autoscale downwards without risk of progression loss for players. For our next project having this fixed would be extremely important. 

I am open to trying to fix it myself, verifying the fix under load and opening a pull request, but for that it would be great to get any advice on where to look (such as the tidbit @benjaminpetit posted above)

**Comment by lpyvvvvvv** [comment]
I have the same problem.
Have you solved the problem yet

**Comment by benjaminpetit** [comment]
We did some heavy changes in 3.x regarding directory/shutdown. @lpyvvvvvv, which version are you using?

Now the only issue if you are using the distributed directory is that `OnDeactivateAsync()` might be running while another activation is being created, but it should not run new requests. We don't recommend doing storage operation during this for several reasons, including this.

**Comment by mikkoairaksinen** [comment]
@benjaminpetit I am a little bit unsure about what you are referring to when you say not recommending storage operations. Do you mean you should not run storage operations during OnDeactivateAsync() ? Or during activation creation? 

If you mean during OnDeactivateAsync() then what is the pattern that should be used to make sure state is never lost in case of a deactivation?

**Comment by benjaminpetit** [comment]
You should not run storage operations during `OnDeactivateAsync()` since it might not be run if the silo shutdown ungracefully (process crash or killed by other silos).

If you really need "strong" single activations, you could implement a lease/lock system, or look at using another grain directory (for example the [Redis Grain Directory](https://www.nuget.org/packages/Microsoft.Orleans.GrainDirectory.Redis/3.3.0-beta1) should not have the same issue.

**Comment by lpyvvvvvv** [comment]
@benjaminpetit  RedisGrainDirectory is useful,thanks a lot

**Comment by mikkoairaksinen** [comment]
@benjaminpetit This is where my confusion comes from, as the reason why actor frameworks are considered to be perfectly suited for gaming backends is that in addition to single thread guarantees they are essentially a very efficient, isolated in-memory caches. For gaming workloads the pattern of 

1) hydrate from database when player session starts
2) perform a lot of state changes (possibly multiple changes per second) in memory
3) flush the state to database once the player is no longer playing and sending state changes

is particularly attractive as normally the database is a huge bottleneck. If saving state on deactivation is not considered a good practice I don't really know how the grain is supposed to have any guarantee of not losing some set of changes during normal operation. We could (and already have) implement a timer-based interval writing to the database to mitigate the loss of state when an ungraceful shutdown occurs, and as long as that timer interval is shorter than the deactivation idle timer then the state should remain consistent in normal operation. However, in the case of a graceful shutdown (autoscaling event for example) an active grain might be shut down and just lose state.

I understand the constraint that an ungraceful shutdown causes loss of state and that is a tradeoff we are willing to make. My problem is I just don't understand how a graceful shutdown could be handled if we can't write state on DeactivateAsync. If not losing state during graceful shutdown is what you mean by "strong" single state activation and using a different grain directory helps, then that sounds good and would solve the problem for us.

**Comment by vrecluse** [comment]
> @benjaminpetit This is where my confusion comes from, as the reason why actor frameworks are considered to be perfectly suited for gaming backends is that in addition to single thread guarantees they are essentially a very efficient, isolated in-memory caches. For gaming workloads the pattern of
> 
> 1. hydrate from database when player session starts
> 2. perform a lot of state changes (possibly multiple changes per second) in memory
> 3. flush the state to database once the player is no longer playing and sending state changes
> 
> is particularly attractive as normally the database is a huge bottleneck. If saving state on deactivation is not considered a good practice I don't really know how the grain is supposed to have any guarantee of not losing some set of changes during normal operation. We could (and already have) implement a timer-based interval writing to the database to mitigate the loss of state when an ungraceful shutdown occurs, and as long as that timer interval is shorter than the deactivation idle timer then the state should remain consistent in normal operation. However, in the case of a graceful shutdown (autoscaling event for example) an active grain might be shut down and just lose state.
> 
> I understand the constraint that an ungraceful shutdown causes loss of state and that is a tradeoff we are willing to make. My problem is I just don't understand how a graceful shutdown could be handled if we can't write state on DeactivateAsync. If not losing state during graceful shutdown is what you mean by "strong" single state activation and using a different grain directory helps, then that sounds good and would solve the problem for us.

We're using Orleans this way, too.
Any best practice or recommended way for this scenario?
@benjaminpetit
